File: kotlin/src/main/kotlin/329.LongestIncreasingPathinaMatrix.kt
import org.junit.Assert.assertEquals
import org.junit.Test
//https://leetcode.com/problems/longest-increasing-path-in-a-matrix/


class LongestIncreasingPathinaMatrix{
  fun longestIncreasingPath(matrix: Array<IntArray>): Int {
    val n = matrix.size
    if(n == 0) return 0
    val m = matrix[0].size
    val cache = Array(n){IntArray(m)}

    fun dp(i: Int, j: Int): Int{
      if(cache[i][j] > 0) return cache[i][j]
      cache[i][j] = 1
      for((ni, nj) in listOf(i to j+1, i to j-1, i+1 to j, i-1 to j))
        if(ni in 0 until n && nj in 0 until m && matrix[ni][nj] > matrix[i][j])
          cache[i][j] = maxOf(cache[i][j], dp(ni, nj)+1)
      return cache[i][j]
    }

    var mx= 0
    for(i in 0 until n)
      for(j in 0 until m)
        mx = maxOf(mx, dp(i, j))
    return mx
  }

  @Test
  fun test1(){
    assertEquals(4, longestIncreasingPath(arrayOf(
        intArrayOf(9,9,4),
        intArrayOf(6,6,8),
        intArrayOf(2,1,1)
    )))
  }

  @Test
  fun test2(){
    assertEquals(4, longestIncreasingPath(arrayOf(
        intArrayOf(3,4,5),
        intArrayOf(3,2,6),
        intArrayOf(2,2,1)
    )))
  }
}

File: python/329_LongestIncreasingPathinaMatrix.py
from unittest import TestCase
# https://leetcode.com/problems/longest-increasing-path-in-a-matrix/


class LongestIncreasingPathinaMatrix(TestCase):
    def longestIncreasingPath(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: int
        """
        if not matrix: return 0
        n, m = len(matrix), len(matrix[0])
        cache = {}

        def dp(i, j):
            if not cache.get((i, j)):
                cache[(i, j)] = max((dp(ni, nj)+1 for (ni, nj) in ((i, j+1), (i, j-1), (i+1, j), (i-1, j))
                                    if 0 <= ni < n and 0 <= nj < m and matrix[ni][nj] > matrix[i][j]),
                                    default=1)
            return cache[(i, j)]

        return max(dp(i, j) for i in range(n) for j in range(m))

    def test1(self):
        self.assertEqual(4, self.longestIncreasingPath(
            [
                [9, 9, 4],
                [6, 6, 8],
                [2, 1, 1]
            ]
        ))

    def test2(self):
        self.assertEqual(4, self.longestIncreasingPath(
            [
                [3, 4, 5],
                [3, 2, 6],
                [2, 2, 1]
            ]
        ))


